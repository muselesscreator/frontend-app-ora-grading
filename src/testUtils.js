/* eslint-disable import/prefer-default-export */
/**
 * mock components
 * @example mockStructuredComponents('Card', { Card: ['Body']})
 * @param {string[]} components
 * @param {map{string, string[]}} nestedComponents
 */
const mockStructuredComponents = (components = [], nestedComponents = {}) => {
  const mockResults = {};
  components.forEach(
    (component) => {
      const fn = () => component;
      // this is necessary to define arrow function a name.
      // Mock is using the function name to define the method.
      Object.defineProperty(fn, 'name', { value: component });
      mockResults[component] = fn;
    },
  );

  Object.keys(nestedComponents).forEach(component => {
    mockResults[component] = mockResults[component] || {};
    nestedComponents[component].forEach(
      (nestedComponent) => {
        mockResults[component][nestedComponent] = `${component}.${nestedComponent}`;
      },
    );
  });

  return mockResults;
};

/**
 * This method is helper for writing mock for paragon components.
 * Note: This assume paragon has only 2 level coomponent at most.
 * @example mockComponents('Card', 'Card.Body')
 * @param  {...string} args
 */
export const mockComponents = (...args) => {
  const components = [];
  const nestedComponents = {};

  args.forEach((component) => {
    if (component.includes('.')) {
      const [name, value] = component.split('.');
      nestedComponents[name] = nestedComponents[name] || [];
      nestedComponents[name].push(value);
    } else {
      components.push(component);
    }
  });

  return mockStructuredComponents(components, nestedComponents);
};

/**
 * mapping = {
 *   Button: 'Button',
 *   DataTable: {
 *     Body: 'DataTable.Body',
 *   }
 * }
 */
export const mockComponents2 = (mapping) => {
  const mockedModule = {};
  Object.keys(mapping).forEach(name => {
    const isNested = typeof mapping[name] === 'object';
    const value = isNested ? name : mapping[name];
    mockedModule[name] = () => value;
    Object.defineProperty(mockedModule[name], 'name', { value });
    if (isNested) {
      Object.keys(mapping[name]).forEach(subComponentName => {
        mockedModule[name][subComponentName] = mapping[name][subComponentName];
      });
    }
  });
  return mockedModule;
};

/**
 * mockNestedComponent('Card', { Body: 'Card.Body', ... });
 */
export const mockNestedComponent = (name, nestedComponents) => {
  const fn = () => name;
  Object.defineProperty(fn, 'name', { value: name });
  Object.keys(nestedComponents).forEach(nestedName => {
    fn[nestedName] = nestedComponents[nestedName];
  });
  return fn;
};
