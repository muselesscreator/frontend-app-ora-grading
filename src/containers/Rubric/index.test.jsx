import React from 'react';
import { shallow } from 'enzyme';

import { formatMessage } from 'testUtils';
import { selectors, thunkActions } from 'data/redux';
import { RequestKeys } from 'data/constants/requests';
import CriterionContainer from 'containers/CriterionContainer';
import { Rubric, mapStateToProps, mapDispatchToProps } from '.';

jest.mock('containers/CriterionContainer', () => 'CriterionContainer');
jest.mock('./RubricFeedback', () => 'RubricFeedback');

jest.mock('data/redux', () => ({
  selectors: {
    app: {
      rubric: {
        criteriaIndices: jest.fn((...args) => ({
          rubricCriteriaIndices: args,
        })),
      },
    },
    grading: {
      selected: {
        isGrading: jest.fn((...args) => ({ isGrading: args })),
      },
    },
    requests: {
      isCompleted: jest.fn((...args) => ({ isCompleted: args })),
      isPending: jest.fn((...args) => ({ isPending: args })),
    },
  },
  thunkActions: {
    grading: {
      submitGrade: jest.fn(),
    },
  },
}));

describe('Rubric Container', () => {
  const props = {
    intl: { formatMessage },
    isCompleted: false,
    gradeIsPending: false,
    lockIsPending: false,
    isGrading: true,
    criteriaIndices: [1, 2, 3, 4, 5],
    submitGrade: jest.fn().mockName('this.props.submitGrade'),
  };
  let el;
  beforeEach(() => {
    el = shallow(<Rubric {...props} />);
    el.instance().submitGradeHandler = jest
      .fn()
      .mockName('this.submitGradeHandler');
  });
  describe('snapshot', () => {
    beforeEach(() => {
      el.instance().submitGradeHandler = jest.fn().mockName('this.submitGradeHandler');
      el.instance().hideDemoAlert = jest.fn().mockName('this.hideDemoAlert');
      jest.spyOn(el.instance(), 'criteria', 'get').mockReturnValue('// get this.criteria()');
    });
    test('is grading', () => {
      expect(el.instance().render()).toMatchSnapshot();
    });
    test('is not grading', () => {
      el.setProps({ isGrading: false });
      expect(el.instance().render()).toMatchSnapshot();
    });
    test('is grading, submit pending', () => {
      el.setProps({ gradeIsPending: true });
      expect(el.instance().render()).toMatchSnapshot();
    });
    test('is grading, lock is pending', () => {
      el.setProps({ lockIsPending: true });
      expect(el.instance().render()).toMatchSnapshot();
    });
    test('submit completed', () => {
      el.setProps({ isCompleted: true, isGrading: false });
      expect(el.instance().render()).toMatchSnapshot();
    });
  });

  describe('component', () => {
    describe('render', () => {
      test('criteria getter returns a CriterionContainer for each criteriaIndices value', () => {
        const container = (value) => (
          shallow(<CriterionContainer isGrading key={value} orderNum={value} />)
        );
        expect(el.instance().criteria).toMatchObject(props.criteriaIndices.map(container));
      });
      test('is grading (grading footer present)', () => {
        expect(el.find('.grading-rubric-footer').length).toEqual(1);
        const containers = el.find('CriterionContainer');
        expect(containers.length).toEqual(props.criteriaIndices.length);
        containers.forEach((container, i) => {
          expect(container.key()).toEqual(String(props.criteriaIndices[i]));
        });
      });

      test('is not grading (no grading footer)', () => {
        el.setProps({
          isGrading: false,
        });
        expect(el.find('.grading-rubric-footer').length).toEqual(0);
        const containers = el.find('CriterionContainer');
        expect(containers.length).toEqual(props.criteriaIndices.length);
        containers.forEach((container, i) => {
          expect(container.key()).toEqual(String(props.criteriaIndices[i]));
        });
      });
    });

    describe('behavior', () => {
      test('submitGrade', () => {
        el = shallow(<Rubric {...props} />);
        el.instance().submitGradeHandler();
        expect(props.submitGrade).toBeCalledTimes(1);
      });
    });
  });

  describe('mapStateToProps', () => {
    const testState = { abitaryState: 'some data' };
    let mapped;
    beforeEach(() => {
      mapped = mapStateToProps(testState);
    });
    test('isGrading from selectors.grading.selected.isGrading', () => {
      expect(mapped.isGrading).toEqual(selectors.grading.selected.isGrading(testState));
    });
    test('gradeIsPending from selectors.requests.isPending(submitGrade)', () => {
      expect(mapped.gradeIsPending).toEqual(
        selectors.requests.isPending(testState, { requestKey: RequestKeys.submitGrade }),
      );
    });
    test('lockIsPending from selectors.requests.isPending(setLock)', () => {
      expect(mapped.lockIsPending).toEqual(selectors.requests.isPending(
        testState, { requestKey: RequestKeys.setLock },
      ));
    });
    test('criteriaIndices from selectors.app.rubric.criteriaIndices', () => {
      expect(mapped.criteriaIndices).toEqual(
        selectors.app.rubric.criteriaIndices(testState),
      );
    });
  });
  describe('mapDispatchToProps', () => {
    beforeEach(() => {});
    test('maps thunkActions.grading.submitGrade to submitGrade prop', () => {
      expect(mapDispatchToProps.submitGrade).toEqual(
        thunkActions.grading.submitGrade,
      );
    });
  });
});
