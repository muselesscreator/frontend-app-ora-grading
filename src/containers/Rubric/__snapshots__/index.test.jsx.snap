// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rubric Container snapshot is grading 1`] = `
<React.Fragment>
  <Card
    className="grading-rubric-card"
  >
    <Card.Body
      className="grading-rubric-body"
    >
      <h3>
        Rubric
      </h3>
      <hr
        className="m-2.5"
      />
      // get this.criteria()
      <hr />
      <RubricFeedback />
    </Card.Body>
    <div
      className="grading-rubric-footer"
    >
      <StatefulButton
        disabledStates={
          Array [
            "pending",
            "complete",
          ]
        }
        labels={
          Object {
            "complete": "Grade Submitted",
            "default": "Submit grade",
            "pending": "Submitting grade",
          }
        }
        onClick={[MockFunction this.submitGradeHandler]}
        state="default"
      />
    </div>
  </Card>
  <DemoAlert
    isOpen={false}
    onClose={[MockFunction this.hideDemoAlert]}
  />
</React.Fragment>
`;

exports[`Rubric Container snapshot is grading, lock is pending 1`] = `
<React.Fragment>
  <Card
    className="grading-rubric-card"
  >
    <Card.Body
      className="grading-rubric-body"
    >
      <h3>
        Rubric
      </h3>
      <hr
        className="m-2.5"
      />
      // get this.criteria()
      <hr />
      <RubricFeedback />
    </Card.Body>
    <div
      className="grading-rubric-footer"
    >
      <StatefulButton
        disabledStates={
          Array [
            "pending",
            "complete",
          ]
        }
        labels={
          Object {
            "complete": "Grade Submitted",
            "default": "Submit grade",
            "pending": "Submitting grade",
          }
        }
        onClick={[MockFunction this.submitGradeHandler]}
        state="pending"
      />
    </div>
  </Card>
  <DemoAlert
    isOpen={false}
    onClose={[MockFunction this.hideDemoAlert]}
  />
</React.Fragment>
`;

exports[`Rubric Container snapshot is grading, submit pending 1`] = `
<React.Fragment>
  <Card
    className="grading-rubric-card"
  >
    <Card.Body
      className="grading-rubric-body"
    >
      <h3>
        Rubric
      </h3>
      <hr
        className="m-2.5"
      />
      // get this.criteria()
      <hr />
      <RubricFeedback />
    </Card.Body>
    <div
      className="grading-rubric-footer"
    >
      <StatefulButton
        disabledStates={
          Array [
            "pending",
            "complete",
          ]
        }
        labels={
          Object {
            "complete": "Grade Submitted",
            "default": "Submit grade",
            "pending": "Submitting grade",
          }
        }
        onClick={[MockFunction this.submitGradeHandler]}
        state="pending"
      />
    </div>
  </Card>
  <DemoAlert
    isOpen={false}
    onClose={[MockFunction this.hideDemoAlert]}
  />
</React.Fragment>
`;

exports[`Rubric Container snapshot is not grading 1`] = `
<React.Fragment>
  <Card
    className="grading-rubric-card"
  >
    <Card.Body
      className="grading-rubric-body"
    >
      <h3>
        Rubric
      </h3>
      <hr
        className="m-2.5"
      />
      // get this.criteria()
      <hr />
      <RubricFeedback />
    </Card.Body>
  </Card>
  <DemoAlert
    isOpen={false}
    onClose={[MockFunction this.hideDemoAlert]}
  />
</React.Fragment>
`;

exports[`Rubric Container snapshot submit completed 1`] = `
<React.Fragment>
  <Card
    className="grading-rubric-card"
  >
    <Card.Body
      className="grading-rubric-body"
    >
      <h3>
        Rubric
      </h3>
      <hr
        className="m-2.5"
      />
      // get this.criteria()
      <hr />
      <RubricFeedback />
    </Card.Body>
    <div
      className="grading-rubric-footer"
    >
      <StatefulButton
        disabledStates={
          Array [
            "pending",
            "complete",
          ]
        }
        labels={
          Object {
            "complete": "Grade Submitted",
            "default": "Submit grade",
            "pending": "Submitting grade",
          }
        }
        onClick={[MockFunction this.submitGradeHandler]}
        state="complete"
      />
    </div>
  </Card>
  <DemoAlert
    isOpen={false}
    onClose={[MockFunction this.hideDemoAlert]}
  />
</React.Fragment>
`;
